File: scratch/analyze_3.js
const fs = require('fs');

const filePath = 'c:\\Users\\72671\\Desktop\\שפות תכנות\\סיכום\\3.html';
const html = fs.readFileSync(filePath, 'utf8');
const lines = html.split('\n');

console.log(`Total lines: ${lines.length}`);

let currentTab = null;
let inScript = false;
let openDivs = 0; // global open divs
let tabList = [];
let globalErrors = [];

for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineNum = i + 1;

    // Check script boundary
    if (line.includes('<script>') || line.includes('<script ')) {
        inScript = true;
    }
    if (inScript) {
        if (line.includes('</script>')) {
            inScript = false;
        }
        continue;
    }

    // Ignore comment-only lines
    if (line.trim().startsWith('<!--') && line.trim().endsWith('-->')) {
        continue;
    }

    // Find tab content starts
    // In 3.html, let's see how tabs are declared. Usually they have class="tab-content" or class="..." and id="..."
    // Let's match `<div id="..."` that represents a tab content
    // Let's look for `<div` with id and check if it's a tab container
    const isTabStart = line.includes('tab-content') && line.includes('<div');
    
    // We also count div tokens in the line
    const divRegex = /<\/?div\b[^>]*>/gi;
    let match;
    let lineDivs = [];
    while ((match = divRegex.exec(line)) !== null) {
        lineDivs.push({
            tag: match[0],
            isClose: match[0].startsWith('</')
        });
    }

    if (isTabStart) {
        const idMatch = line.match(/id="([^"]+)"/);
        const tabId = idMatch ? idMatch[1] : 'unknown';
        
        if (currentTab) {
            currentTab.endLine = lineNum - 1;
            tabList.push(currentTab);
        }

        currentTab = {
            id: tabId,
            startLine: lineNum,
            endLine: null,
            openCount: 0,
            closeCount: 0,
            depthStack: [],
            startGlobalDivs: openDivs
        };
    }

    lineDivs.forEach(div => {
        if (div.isClose) {
            openDivs--;
            if (currentTab) {
                currentTab.closeCount++;
                if (currentTab.depthStack.length > 0) {
                    currentTab.depthStack.pop();
                } else {
                    // Underflow: closed the tab wrapper itself, or closed more than opened inside
                    // Let's record this
                }
            }
        } else {
            openDivs++;
            if (currentTab) {
                currentTab.openCount++;
                const idAttr = div.tag.match(/id="([^"]+)"/);
                const classAttr = div.tag.match(/class="([^"]+)"/);
                currentTab.depthStack.push({
                    line: lineNum,
                    tag: div.tag,
                    id: idAttr ? idAttr[1] : null,
                    class: classAttr ? classAttr[1] : null
                });
            }
        }
    });

    if (currentTab && openDivs <= currentTab.startGlobalDivs) {
        // If we drop below or hit the global div count we started with, the tab has ended!
        currentTab.endLine = lineNum;
        tabList.push(currentTab);
        currentTab = null;
    }
}

if (currentTab) {
    currentTab.endLine = lines.length;
    tabList.push(currentTab);
}

console.log("\n=== TABS DETECTED IN 3.HTML ===");
tabList.forEach(t => {
    console.log(`Tab: "${t.id}" (Lines ${t.startLine} -> ${t.endLine})`);
    console.log(`  Opened divs: ${t.openCount}, Closed divs: ${t.closeCount}`);
    console.log(`  Unclosed at end of tab: ${t.depthStack.length}`);
    if (t.depthStack.length > 0) {
        console.log(`  Unclosed list:`);
        t.depthStack.forEach(d => {
            console.log(`    Line ${d.line}: ${d.tag}`);
        });
    }
});

console.log(`\nFinal global open divs in document: ${openDivs}`);

File: scratch/analyze_3.py
import re

file_path = r'c:\Users\72671\Desktop\שפות תכנות\סיכום\3.html'
with open(file_path, 'r', encoding='utf-8') as f:
    lines = f.readlines()

print(f"Total lines in 3.html: {len(lines)}")

# 1. Let's find all lines that start or contain tab-content divs
tab_regex = re.compile(r'class="[^"]*tab-content[^"]*"')
tab_id_regex = re.compile(r'id="([^"]+)"')

tabs = []
current_tab = None

# Let's count open/close divs in script tags to ignore them
in_script = False

for idx, line in enumerate(lines):
    line_num = idx + 1
    
    if '<script>' in line or '<script ' in line:
        in_script = True
        continue
    if '</script>' in line:
        in_script = False
        continue
        
    if in_script:
        continue

    # Search for tab content start
    # Let's see if the line matches tab-content
    if 'tab-content' in line and '<div' in line:
        id_match = tab_id_regex.search(line)
        tab_id = id_match.group(1) if id_match else 'unknown'
        
        if current_tab:
            tabs.append(current_tab)
        current_tab = {
            'id': tab_id,
            'start_line': line_num,
            'div_stack': [],
            'open_divs': 0,
            'close_divs': 0,
            'net_balance': 0,
            'end_line': None
        }
        
    if current_tab:
        # We count div occurrences
        open_tags = len(re.findall(r'<div\b', line))
        close_tags = len(re.findall(r'</div>', line))
        current_tab['open_divs'] += open_tags
        current_tab['close_divs'] += close_tags
        current_tab['net_balance'] += (open_tags - close_tags)
        
        # We can also track tag stack within tab
        # (Though we might have open div from the tab wrapper itself, so balance should end at 0 or 1 depending on where we close it)
        # If net_balance drops to 0 (which means the wrapper div itself is closed), that might be the end of the tab
        if current_tab['net_balance'] <= 0 and (open_tags > 0 or close_tags > 0):
            # This tab might have been closed!
            current_tab['end_line'] = line_num
            tabs.append(current_tab)
            current_tab = None

if current_tab:
    tabs.append(current_tab)

print("\n--- Tabs Found and Balances ---")
for t in tabs:
    print(f"Tab ID: {t['id']}")
    print(f"  Line range: {t['start_line']} to {t['end_line']}")
    print(f"  Total open divs inside/including tab container: {t['open_divs']}")
    print(f"  Total close divs: {t['close_divs']}")
    print(f"  Net balance at end: {t['net_balance']}")
